src/Workspaces/Core/Portable/Classification/SyntaxClassification/SyntacticChangeRangeComputer.cs

```diff
@@ -0,0 +1,249 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Classification
+{
+    /// <summary>
+    /// Computes a syntactic text change range that determines the range of a document that was changed by an edit. The
+    /// portions outside this change range are guaranteed to be syntactically identical (see <see
+    /// cref="SyntaxNode.IsIncrementallyIdenticalTo"/>).  This algorithm is intended to be <em>fast</em>.  It is
+    /// technically linear in the number of nodes and tokens that may need to examined.  However, in practice, it should
+    /// operate in sub-linear time as it will bail the moment tokens don't match, and it's able to skip over matching
+    /// nodes fully without examining the contents of those nodes.  This is intended for consumers that want a
+    /// reasonably accurate change range computer, but do not want to spend an inordinate amount of time getting the
+    /// most accurate and minimal result possible.
+    /// </summary>
+    /// <remarks>
+    /// This computation is not guaranteed to be minimal.  It may return a range that includes parts that are unchanged.
+    /// This means it is also legal for the change range to just specify the entire file was changed. The quality of
+    /// results will depend on how well the parsers did with incremental parsing, and how much time is given to do the
+    /// comparison.  In practice, for large files (i.e. 15kloc) with standard types of edits, this generally returns
+    /// results in around 50-100 usecs on a i7 3GHz desktop.
+    /// <para>
+    /// This algorithm will respect the timeout provided to the best of abilities.  If any information has been computed
+    /// when the timeout elapses, it will be returned.
+    /// </para>
+    /// </remarks>
+    internal static class SyntacticChangeRangeComputer
+    {
+        private static readonly ObjectPool<Stack<SyntaxNodeOrToken>> s_pool = new(() => new());
+
+        public static async ValueTask<TextChangeRange?> ComputeSyntacticChangeRangeAsync(
```

Comment:
It's unclear why we allow returning null here. It looks like cases where we timeout waiting on the syntax roots. We still had to compute them unless the cancellation token cancelled, can we use that to return a non-null value here?

Reply:
Not exactly.  We may not have canceled, but we may have otherwise exceeded out time budget.  "Cancellation" in these codepaths is for saying: no, stop everything, all results are useless as the client has moved on.

The diffTimeout is different.  It's: we still need the results, but we need the fast, so don't do superflous work.

Does that make sense?

Reply:
Yea, I meant that exceeding the time budge can return null if the root wasn't retrieved in time. It seems odd we have null in that case, sense we're saying "Assume both of these are completely different",  based on the comment. An actual result with the whole tree being different would be easier to understand and consume than guessing what null meant. Is it no difference? Is it all different?

Reply:
> An actual result with the whole tree being different would be easier to understand and consume than guessing what null meant. Is it no difference? Is it all different?

I see what you mean.  To do this, we'd still need the sizes of both the old and new tree.  And that means we'd have to then get both roots.  I found it easier to just return `null` trivially and have the caller know that means that the narrow range could not be computed :)